### PR TITLE
REF/CLN: pandas/io/parsers.py

### PR DESCRIPTION
**Pull request by ivanovmg**
- [ ] closes #xxxx
- [ ] tests added / passed
- [ ] passes `black pandas`
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

Refactor/cleanup ``pandas/io/parsers.py``
- Extract method _refresh_kwargs_based_on_dialect
- Extract method _validate_skipfooter
- Drop local variable engine_specified
- Clean-up FutureWarning issue

### CODE REVIEW

**Comment by jbrockmendel** [comment]
big picture: whats the goal?

File: pandas/io/parsers.py [hunk 1]
```diff
@@ -1106,25 +1112,18 @@ def _clean_options(self, options, engine):
...
-
-            msg = (
-                f"The {repr(arg)} argument has been deprecated and will be "
-                "removed in a future version."
-            )
-
             if result.get(arg, depr_default) != depr_default:
-                depr_warning += msg + "\n\n"
+                msg = (
+                    f"The {repr(arg)} argument has been deprecated and will be "
```
Comment (jbrockmendel) [inline pandas/io/parsers.py:1120]: isnt `repr` here redundant?

**Comment by jbrockmendel** [comment]
> Extract some methods to make any further refactoring easier.

Is there further refactoring you have in mind?

File: pandas/io/parsers.py [hunk 2]
```diff
@@ -951,6 +903,61 @@ def __init__(self, f, engine=None, **kwds):
     def close(self):
         self._engine.close()
 
+    @staticmethod
```
Comment (jreback) [inline pandas/io/parsers.py:906]: don't use staticmethods

File: pandas/io/parsers.py [hunk 3]
```diff
@@ -951,6 +903,61 @@ def __init__(self, f, engine=None, **kwds):
...
+                # Refer to the outer "_read_" function for more info.
+                if not (param == "delimiter" and kwds.pop("sep_override", False)):
+                    conflict_msgs.append(msg)
+
+            if conflict_msgs:
+                warnings.warn("\n\n".join(conflict_msgs), ParserWarning, stacklevel=2)
+            kwds[param] = dialect_val
+        return kwds
+
+    @staticmethod
```
Comment (jreback) [inline pandas/io/parsers.py:953]: same

File: pandas/io/parsers.py [hunk 4]
```diff
@@ -951,6 +903,61 @@ def __init__(self, f, engine=None, **kwds):
...
+    def _validate_skipfooter(kwds):
```
Comment (jreback) [inline pandas/io/parsers.py:954]: can you type

File: pandas/io/parsers.py [hunk 5]
```diff
@@ -951,6 +903,61 @@ def __init__(self, f, engine=None, **kwds):
     def close(self):
         self._engine.close()
 
+    @staticmethod
+    def _refresh_kwargs_based_on_dialect(kwds, dialect):
```
Comment (jreback) [inline pandas/io/parsers.py:907]: can you type and add a doc-string

File: pandas/io/parsers.py [hunk 6]
```diff
@@ -867,71 +867,23 @@ def __init__(self, f, engine=None, **kwds):
...
             engine_specified = False
-
+        self.engine = engine
         self._engine_specified = kwds.get("engine_specified", engine_specified)
 
         if kwds.get("dialect") is not None:
             dialect = kwds["dialect"]
             if dialect in csv.list_dialects():
                 dialect = csv.get_dialect(dialect)
+            kwds = self._refresh_kwargs_based_on_dialect(kwds, dialect)
```
Comment (jreback) [inline pandas/io/parsers.py:877]: merge master as some of this updated; and rename this like check_defaults_dialect (or similar)

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/io/parsers.py (hunk 5 above)
Comment (ivanovmg) [reply pandas/io/parsers.py]: Done

File: pandas/io/parsers.py (hunk 4 above)
Comment (ivanovmg) [reply pandas/io/parsers.py]: Done

File: pandas/io/parsers.py (hunk 2 above)
Comment (ivanovmg) [reply pandas/io/parsers.py]: Moved to module level

File: pandas/io/parsers.py [hunk 7]
```diff
@@ -3791,3 +3735,93 @@ def _check_defaults_read(
         kwds["engine_specified"] = False
 
     return kwds
+
+
+def _merge_with_dialect_properties(
```
Comment (jreback) [inline pandas/io/parsers.py:3740]: reverse these (always put kwds things last).

Comment (jreback) [reply pandas/io/parsers.py:3740]: rename kwds -> defaults

so similar to the _check_defaults_read

also not averse to renaming _check_defaults_read to something better.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/io/parsers.py (hunk 7 above)
Comment (ivanovmg) [reply pandas/io/parsers.py:3742]: I have done it.
Also, I found the issue with the ValueError message, when dialect is incorrect.
Originally, kwds['dialect'], but I decided to use dialect.__name__ for better readability, which is not accepted by mypy (throws error ``"Dialect" has no attribute "__name__"  [attr-defined]``).
So, now I use string representation of dialect instance.

**Review by jreback — APPROVED** [review]
lgtm. cc @gfyoung if any comments, pls merge when ready.

**Comment by jreback** [comment]
cc @twoertwein if any comments.

**Comment by twoertwein** [comment]
LGTM, outsourcing the dialect validation makes the constructor of `TextFileReader` easier to read.

**Review by gfyoung — APPROVED** [review]

**Comment by gfyoung** [comment]
Thanks @ivanovmg!

**Pull request merged**